src/main/java/org/jenkinsci/plugins/registry/notification/events/EventType.java

```diff
@@ -0,0 +1,68 @@
+package org.jenkinsci.plugins.registry.notification.events;
+
+import hudson.EnvVars;
+import hudson.model.AbstractDescribableImpl;
+import net.sf.json.JSONObject;
+import org.jenkinsci.plugins.registry.notification.webhook.PushNotification;
+import org.jenkinsci.plugins.registry.notification.webhook.WebHookPayload;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+
+import javax.annotation.Nonnull;
+import java.text.ParseException;
+import java.util.logging.Logger;
+
+public abstract class EventType extends AbstractDescribableImpl<EventType> {
+    private static final Logger logger = Logger.getLogger(EventType.class.getName());
+    public static final String KEY_DOCKER_IMAGE_DIGEST = WebHookPayload.PREFIX + "DOCKER_IMAGE_DIGEST";
+
+    public boolean accepts(@Nonnull String jsonTypeCode ){
+        return jsonTypeCode.equals(getType());
+    }
+
+    public void buildEnvironment(@Nonnull EnvVars envs, @Nonnull PushNotification push) {
+        // Each event type can have a unique time stamp field name
+        WebHookPayload payload = push.getWebHookPayload();
+        if( payload != null ){
+            JSONObject data = payload.getData();
+            if ( data != null) {
+                JSONObject contents = data.getJSONObject("contents");
+                if (contents != null) {
+                    DateTimeFormatter parser = ISODateTimeFormat.dateTimeParser();
+                    String timeStamp = getTimeStamp(contents);
+                    if( timeStamp != null ) {
+                        push.setPushedAt(parser.parseDateTime(timeStamp).toDate());
```

Comment:
Yea this is still weird, having the data being mutated here. It's a bit out of the blue and quite unexpected that `buildEnvironment` would mutate the data at all.
The timestamp isn't even used in the environment as far as I can see, so double wrong to put it here.
`pushedAt` should really be set in the constructor.
In the test data I can see two timestamps I guess one for the event itself and one for when the thing happened to the image. The one for the event itself seems consistent and I guess also matches the name `pushedAt`, or maybe not.
Anyway this should be in the constructor and not here :)

Reply:
@rsandell Going to dust this thing off and try to get it done , are you around to do reviews of any changes I make this week?

Reply:
I am around for about three more weeks before vacation, but I'm fully focused on a new project so not noticing review pings as well as I should _(as you might notice by me responding 9 days later)_. So ping me via some more direct channel and I can reserve some time.